pyfr/integrators/writers.py

```diff
@@ -60,44 +110,34 @@ def _get_name_for_soln(self, etype, prank=None):
         prank = prank or self.rallocs.prank
         return 'soln_{}_p{}'.format(etype, prank)
 
-
-class FileWriter(BaseWriter):
-    writer_name = 'pyfrs-file'
-
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-
-        # MPI info
+    def _write_parallel(self, path, solnmap, metadata):
         comm, rank, root = get_comm_rank_root()
 
-        # Get the type and shape of each element in the partition
-        etypes, shapes = self._system.ele_types, self._system.ele_shapes
-
-        # Gather this information onto the root rank
-        eleinfo = comm.gather(zip(etypes, shapes), root=root)
-
-        if rank == root:
-            self._mpi_rbufs = mpi_rbufs = []
-            self._mpi_rreqs = mpi_rreqs = []
-            self._mpi_names = mpi_names = []
-            self._loc_names = loc_names = []
-
-            for mrank, meleinfo in enumerate(eleinfo):
-                prank = self.rallocs.mprankmap[mrank]
-                for tag, (etype, dims) in enumerate(meleinfo):
-                    name = self._get_name_for_soln(etype, prank)
-
-                    if mrank == root:
-                        loc_names.append(name)
-                    else:
-                        rbuf = np.empty(dims, dtype=self.backend.fpdtype)
-                        rreq = comm.Recv_init(rbuf, mrank, tag)
-
-                        mpi_rbufs.append(rbuf)
-                        mpi_rreqs.append(rreq)
-                        mpi_names.append(name)
-
-    def _write(self, path, solnmap, metadata):
+        with h5py.File(path, 'w', driver='mpio', comm=comm) as h5file:
+            smap = {}
+            for s, shape in self.sollist:
+                smap[s] = h5file.create_dataset(
+                                 s, shape, dtype=self.backend.fpdtype
```

Comment:
Extra indentation.  Should be one level deeper than `smap[s]`.  The file `pyfr.solvers.baseadvec.elements` is a good reference here.